cpp/src/umap/simpl_set_embed/algo.h

```diff
@@ -96,70 +95,70 @@ __device__ __host__ T clip(T val, T lb, T ub) {
 }
 
 /**
-	         * Calculate the repulsive gradient
-	         */
-template <typename T>
-__device__ __host__ T repulsive_grad(T dist_squared, float gamma,
-                                     UMAPParams params) {
-  T grad_coeff = 2.0 * gamma * params.b;
+ * Calculate the repulsive gradient
+ */
+__device__ __host__ double repulsive_grad(double dist_squared, double gamma,
```

Comment:
Another custom kernel.

Reply:
This isn’t a kernel, it’s just a utility function that happens to be a __device__ function.